File: frontend/src/types/extractionMemory.ts
/** ARCH41-S3:types — mirrors app/schemas/extraction_memory.py field for field. */

export type MemoryMode = "OFF" | "SHADOW" | "AUTO";
export type TemplateState = "LEARNING" | "TRIAL" | "ACTIVE" | "REJECTED";
export type RuleState = "SHADOW" | "ACTIVE" | "RETIRED";
export type TrialState = "RUNNING" | "PROMOTED" | "REJECTED" | "ABANDONED";

export interface MemorySettings {
  readonly mode: MemoryMode;
  readonly is_default: boolean;
  readonly updated_at?: string | null;
}

export interface MemoryPotential {
  readonly reviewed_documents: number;
  readonly corrected_fields: number;
}

export interface MemorySummary {
  readonly mode: MemoryMode;
  readonly templates: number;
  readonly active_templates: number;
  readonly exemplar_documents: number;
  readonly corrected_exemplars: number;
  readonly active_rules: number;
  readonly shadow_rules: number;
  readonly running_trials: number;
  readonly baseline_correction_rate?: number | null;
  readonly memory_correction_rate?: number | null;
}

export interface FieldMetric {
  readonly field_path: string;
  readonly corrections: number;
  readonly confirmations: number;
  readonly baseline_rate?: number | null;
  readonly memory_rate?: number | null;
}

export interface TemplateRow {
  readonly id: string;
  readonly document_type: string;
  readonly state: TemplateState;
  readonly member_count: number;
  readonly exemplar_documents: number;
  readonly anchor_tokens: readonly string[];
  readonly activated_at?: string | null;
  readonly baseline_correction_rate?: number | null;
  readonly memory_correction_rate?: number | null;
  readonly fields: readonly FieldMetric[];
}

export interface RuleRow {
  readonly id: string;
  readonly template_id: string;
  readonly field_path: string;
  readonly anchor_norm: string;
  readonly offset_dx: number;
  readonly offset_dy: number;
  readonly value_token_count: number;
  readonly support: number;
  readonly replay_hits: number;
  readonly replay_total: number;
  readonly wilson_lower?: number | null;
  readonly live_hits: number;
  readonly live_total: number;
  readonly state: RuleState;
  readonly activated_at?: string | null;
  readonly retired_reason?: string | null;
  readonly sentence: string;
}

export interface TrialRow {
  readonly id: string;
  readonly template_id: string;
  readonly document_type: string;
  readonly state: TrialState;
  readonly on_docs: number;
  readonly off_docs: number;
  readonly on_correction_rate?: number | null;
  readonly off_correction_rate?: number | null;
  readonly p_value?: number | null;
  readonly decision_reason?: string | null;
  readonly started_at: string;
  readonly decided_at?: string | null;
  readonly sentence: string;
}

export interface LearnedField {
  readonly field_path: string;
  readonly corrections: number;
  readonly confirmations: number;
  readonly sentence: string;
  readonly anchor_state?: string | null;
}

export interface WorkItemMemory {
  readonly applied: boolean;
  readonly arm?: string | null;
  readonly injected: boolean;
  readonly template_id?: string | null;
  readonly document_type?: string | null;
  readonly template_state?: string | null;
  readonly layout_documents: number;
  readonly exemplar_documents_used: number;
  readonly fields: readonly LearnedField[];
  readonly headline: string;
}

File: backend/app/schemas/extraction_memory.py
"""ARCH41-S2:schemas — extraction memory API shapes."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class MemorySettingsResponse(BaseModel):
    mode: Literal["OFF", "SHADOW", "AUTO"]
    is_default: bool
    updated_at: Optional[datetime] = None


class MemorySettingsUpdate(BaseModel):
    mode: Literal["OFF", "SHADOW", "AUTO"]


class MemoryPotentialResponse(BaseModel):
    reviewed_documents: int
    corrected_fields: int


class MemorySummaryResponse(BaseModel):
    mode: Literal["OFF", "SHADOW", "AUTO"]
    templates: int
    active_templates: int
    exemplar_documents: int
    corrected_exemplars: int
    active_rules: int
    shadow_rules: int
    running_trials: int
    baseline_correction_rate: Optional[float] = None
    memory_correction_rate: Optional[float] = None


class FieldMetric(BaseModel):
    field_path: str
    corrections: int
    confirmations: int
    baseline_rate: Optional[float] = None
    memory_rate: Optional[float] = None


class TemplateRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_type: str
    state: str
    member_count: int
    exemplar_documents: int
    anchor_tokens: list[str]
    activated_at: Optional[datetime] = None
    baseline_correction_rate: Optional[float] = None
    memory_correction_rate: Optional[float] = None
    fields: list[FieldMetric] = Field(default_factory=list)


class TemplateAction(BaseModel):
    action: Literal["reset"]


class RuleRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    template_id: uuid.UUID
    field_path: str
    anchor_norm: str
    offset_dx: int
    offset_dy: int
    value_token_count: int
    support: int
    replay_hits: int
    replay_total: int
    wilson_lower: Optional[float] = None
    live_hits: int
    live_total: int
    state: str
    activated_at: Optional[datetime] = None
    retired_reason: Optional[str] = None
    sentence: str


class RuleAction(BaseModel):
    action: Literal["retire", "shadow"]


class TrialRow(BaseModel):
    id: uuid.UUID
    template_id: uuid.UUID
    document_type: str
    state: str
    on_docs: int
    off_docs: int
    on_correction_rate: Optional[float] = None
    off_correction_rate: Optional[float] = None
    p_value: Optional[float] = None
    decision_reason: Optional[str] = None
    started_at: datetime
    decided_at: Optional[datetime] = None
    sentence: str


class LearnedField(BaseModel):
    field_path: str
    corrections: int
    confirmations: int
    sentence: str
    anchor_state: Optional[str] = None


class WorkItemMemoryResponse(BaseModel):
    applied: bool
    arm: Optional[str] = None
    injected: bool = False
    template_id: Optional[uuid.UUID] = None
    document_type: Optional[str] = None
    template_state: Optional[str] = None
    layout_documents: int = 0
    exemplar_documents_used: int = 0
    fields: list[LearnedField] = Field(default_factory=list)
    headline: str
